basic/bittwister_test.go

```diff
@@ -0,0 +1,609 @@
+package basic
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"os"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/celestiaorg/knuu/pkg/knuu"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	iperfImage  = "docker.io/clearlinux/iperf:latest"
+	gopingImage = "ghcr.io/celestiaorg/goping:1cc7b33"
+)
+
+func TestBittwister_Bandwidth(t *testing.T) {
+	t.Parallel()
+	// Setup
+
+	const (
+		iperfTestDuration    = 45 * time.Second
+		iperfParallelClients = 4
+		commandTimeout       = 15 * time.Minute
+	)
+
+	iperfMother, err := knuu.NewInstance("iperf")
+	require.NoError(t, err, "Error creating instance")
+
+	require.NoError(t, iperfMother.EnableBitTwister(), "Error enabling BitTwister")
+	err = iperfMother.SetImage(iperfImage)
+	require.NoError(t, err, "Error setting image")
+
+	err = iperfMother.SetCommand("iperf3", "-s")
+	require.NoError(t, err, "Error executing command")
+
+	require.NoError(t, iperfMother.AddPortTCP(5201), "Error adding port")
+	require.NoError(t, iperfMother.Commit(), "Error committing instance")
+
+	iperfServer, err := iperfMother.CloneWithName("iperf-server")
+	require.NoError(t, err, "Error cloning instance")
+
+	iperfClient, err := iperfMother.CloneWithName("iperf-client")
+	require.NoError(t, err, "Error cloning instance")
+
+	t.Cleanup(func() {
+		if os.Getenv("KNUU_SKIP_CLEANUP") == "true" {
+			t.Log("Skipping cleanup")
+			return
+		}
+
+		require.NoError(t, iperfServer.Destroy(), "Error destroying iperf-server instance")
+		require.NoError(t, iperfClient.Destroy(), "Error destroying iperf-client instance")
+	})
+
+	// Prepare iperf client & server
+
+	iperfServerIP, err := iperfServer.GetIP()
+	require.NoError(t, err, "Error getting IP")
+
+	require.NoError(t, iperfServer.EnableBitTwister(), "Error enabling BitTwister")
+	require.NoError(t, iperfServer.Start(), "Error starting iperf-server instance")
+	require.NoError(t, iperfServer.WaitInstanceIsRunning(), "Error waiting for iperf-server instance to be running")
+
+	// Forward the port to the local machine
+	fwdBtPort, err := iperfServer.PortForwardTCP(iperfServer.BitTwister.Port())
+	require.NoError(t, err, "Error port forwarding")
+	iperfServer.BitTwister.SetPort(fwdBtPort)
+	iperfServer.BitTwister.SetNewClientByIPAddr("http://localhost")
```

Comment:
Is there a way to abstract that away so that the user does not need to care about forwarding a port?

Reply:
yes it should be possible

Reply:
yes, fixed